### PR TITLE
feat: propagate trace attributes onto all child spans on update

### PR DESCRIPTION
**Pull request by Steffen911**
> [!IMPORTANT]
> This PR adds context propagation for trace attributes using OpenTelemetry, automating attribute propagation to child spans and deprecating manual trace updates.
> 
>   - **Behavior**:
>     - Introduces context propagation for `userId`, `sessionId`, and `metadata` using OpenTelemetry in `span-processor.ts`.
>     - Deprecates `updateActiveTrace()` in favor of new context propagation methods.
>   - **Functions**:
>     - Adds `withUser`, `withSession`, `withMetadata`, `withContext` in `context.ts` for automatic attribute propagation.
>     - Updates `LangfuseBaseObservation` with instance methods for context propagation.
>   - **Files**:
>     - New `context-keys.ts` for defining OpenTelemetry context keys.
>     - Updates `spanWrapper.ts` to include new context methods.
>     - Comprehensive tests added in `tracing.integration.test.ts` for context propagation scenarios.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=langfuse%2Flangfuse-js&utm_source=github&utm_medium=referral)<sup> for 3411e7dcb81961608e5b5468e85c4dfe70087d3c. You can [customize](https://app.ellipsis.dev/langfuse/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

----

**Disclaimer**: Experimental PR review
---

<h2>Greptile Overview</h2>

Updated On: 2025-10-01 06:42:22 UTC

<h3>Summary</h3>
This PR introduces comprehensive context propagation functionality for Langfuse trace attributes (`userId`, `sessionId`, `metadata`) that automatically flow from parent to child spans without manual intervention. The implementation leverages OpenTelemetry's context and baggage mechanisms to ensure trace-level attributes are consistently applied across all nested operations.

The core changes include:

1. **New Context Keys Infrastructure**: A new `context-keys.ts` file in the core package defines OpenTelemetry context keys (`LANGFUSE_CTX_USER_ID`, `LANGFUSE_CTX_SESSION_ID`, `LANGFUSE_CTX_METADATA`) that serve as the foundation for context propagation across packages.

2. **Context Propagation Functions**: Four new functions (`withUser`, `withSession`, `withMetadata`, `withContext`) in the tracing package enable developers to set trace attributes that automatically propagate to child spans within a callback scope. These functions support both local propagation via OpenTelemetry context and cross-service propagation via baggage.

3. **Instance Methods on Observations**: The observation wrapper classes now include instance methods that wrap the standalone context functions, providing a more fluent API for developers working with observation instances.

4. **Automatic Span Processing**: The OpenTelemetry span processor now automatically reads propagated context values and applies them as span attributes during span creation, ensuring consistent trace metadata across all child spans.

5. **Deprecation Path**: The existing `updateActiveTrace()` method is deprecated in favor of the new context propagation approach, with comprehensive migration guidance provided.

This change fundamentally improves the developer experience by eliminating the need to manually set trace attributes on every span, while following OpenTelemetry best practices for context propagation. The implementation includes proper error handling, type safety, security considerations for cross-service propagation, and comprehensive test coverage.

## Important Files Changed

<details><summary>Changed Files</summary>

| Filename | Score | Overview |
|----------|-------|----------|
| packages/core/src/context-keys.ts | 5/5 | New file defining OpenTelemetry context keys for trace attribute propagation |
| packages/core/src/index.ts | 5/5 | Added export for context-keys module to expose shared constants |
| packages/core/package.json | 4/5 | Added `@opentelemetry/api` as peer dependency for context key creation |
| packages/tracing/src/context.ts | 4/5 | New comprehensive context propagation utilities with proper error handling |
| packages/tracing/src/spanWrapper.ts | 4/5 | Added context propagation instance methods and deprecated updateTrace |
| packages/tracing/src/index.ts | 4/5 | Modified parent context creation and deprecated updateActiveTrace function |
| packages/otel/src/span-processor.ts | 4/5 | Added automatic context attribute propagation during span creation |
| tests/integration/tracing.integration.test.ts | 5/5 | Comprehensive test coverage for all context propagation scenarios |

</details>

## Confidence score: 4/5

- This PR introduces significant new functionality with proper testing but involves complex context propagation logic that could have edge cases
- Score reflects thorough implementation with good error handling and comprehensive tests, but the complexity of context propagation across multiple packages introduces some risk
- Pay close attention to the span processor changes and context propagation logic, especially the metadata flattening and baggage size limits

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant SpanProcessor as "LangfuseSpanProcessor"
    participant Context as "OpenTelemetry Context"
    participant Span as "OpenTelemetry Span"

    User->>SpanProcessor: onStart(span, parentContext)
    Note over SpanProcessor: Add environment and release attributes
    SpanProcessor->>SpanProcessor: propagateContextAttributes(span, parentContext)
    SpanProcessor->>Context: getValue(LANGFUSE_CTX_USER_ID)
    Context-->>SpanProcessor: userId or undefined
    SpanProcessor->>Context: getValue(LANGFUSE_CTX_SESSION_ID)
    Context-->>SpanProcessor: sessionId or undefined
    SpanProcessor->>Context: getValue(LANGFUSE_CTX_METADATA)
    Context-->>SpanProcessor: metadata object or undefined
    
    alt userId exists
        SpanProcessor->>Span: setAttribute(TRACE_USER_ID, userId)
    end
    
    alt sessionId exists
        SpanProcessor->>Span: setAttribute(TRACE_SESSION_ID, sessionId)
    end
    
    alt metadata exists
        loop for each metadata key
            SpanProcessor->>Span: setAttribute("langfuse.metadata.{key}", value)
        end
    end
    
    SpanProcessor->>SpanProcessor: processor.onStart(span, parentContext)
```

### CODE REVIEW

File: packages/tracing/src/context.ts [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+ *   return result;
+ * });
+ *
+ * // Nested contexts (inner overrides outer)
+ * withMetadata({ env: 'prod' }, () => {
+ *   withMetadata({ env: 'staging' }, () => {
+ *     // This scope has env='staging'
+ *   });
+ *   // This scope has env='prod'
+ * });
```
[the remark this answers is not in the document]
Comment (maxdeichmann) [reply packages/tracing/src/context.ts:239]: I think we should have a unified behavior. I guess I am missing some context but cant we first get the context, merge the metadata, and then update it?

**Comment by hassiebp** [comment]
Thanks @Steffen911 - closing this in favor of #676
